File: 1478-maximum-number-of-events-that-can-be-attended/maximum-number-of-events-that-can-be-attended.java
class Solution {
    public int maxEvents(int[][] events) {
        int eventCount = 0;
        int n = events.length;
        Arrays.sort(events,(a,b) -> {
            return a[0] - b[0];
        });
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        int maxDay = 0;
        for(int i=0; i<n; i++) {
            maxDay = Math.max(maxDay,events[i][1]);
        }
        int days=0;
        for(int i=1; days<=maxDay+1; i++) {
            while(pq.size() != 0 && pq.peek() < i) {
                pq.poll();
            }
            while(days < n && events[days][0] == i) {
                pq.offer(events[days][1]);
                days++;
            }
            if(pq.size() != 0 && pq.peek() >= i) {
                pq.poll();
                eventCount++;
            }
            if(pq.size() == 0 && days == n) break;
        }

        return eventCount;
    }
}

File: 1478-maximum-number-of-events-that-can-be-attended/maximum-number-of-events-that-can-be-attended.cpp
class Solution {
public:
    int maxEvents(vector<vector<int>>& arr) {
       int n=arr.size();
       int maxDay=0;
       priority_queue<int,vector<int>,greater<int>> pq;
       for(int i=0;i<n;i++){
           maxDay=max(maxDay,arr[i][1]);
       }
    sort(begin(arr),end(arr));
       int i=0;int ans=0;
       for(int d=1;i<=maxDay+1;d++){
         while(pq.size() && pq.top()<d){
             pq.pop();
         }
         while(i<n && arr[i][0]==d){
             pq.push(arr[i][1]);
             i++;
         }
         if(pq.size() && pq.top()>=d){
             pq.pop();
             ans++;
         }
         if(pq.size()==0 && i==n){
             break;
         }
       }
       return ans;
    }
};